File: api/lib/ingestion/moxfield.ts
import { ParsedDeck, DeckCard } from './to-dck';
import { moxfieldFetch } from '../moxfield-service';

const MOXFIELD_API_BASE = 'https://api2.moxfield.com/v3';

// Moxfield URL patterns:
// https://moxfield.com/decks/ABC123
// https://www.moxfield.com/decks/ABC123
export function isMoxfieldUrl(url: string): boolean {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return false;
    if (parsed.hostname !== 'moxfield.com' && parsed.hostname !== 'www.moxfield.com') return false;
    return /^\/decks\/([a-zA-Z0-9_-]+)/.test(parsed.pathname);
  } catch {
    return false;
  }
}

export function extractMoxfieldDeckId(url: string): string | null {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== 'http:' && parsed.protocol !== 'https:') return null;
    if (parsed.hostname !== 'moxfield.com' && parsed.hostname !== 'www.moxfield.com') return null;
    const match = parsed.pathname.match(/^\/decks\/([a-zA-Z0-9_-]+)/);
    return match ? match[1] : null;
  } catch {
    return null;
  }
}

interface MoxfieldCard {
  quantity: number;
  boardType: string;
  card: {
    name: string;
  };
}

interface MoxfieldBoard {
  count: number;
  cards: Record<string, MoxfieldCard>;
}

interface MoxfieldDeckResponse {
  name: string;
  boards: {
    mainboard?: MoxfieldBoard;
    commanders?: MoxfieldBoard;
    companions?: MoxfieldBoard;
    [key: string]: MoxfieldBoard | undefined;
  };
}

function boardCards(board: MoxfieldBoard | undefined, isCommander: boolean): DeckCard[] {
  if (!board?.cards) return [];
  return Object.values(board.cards).map(entry => ({
    name: entry.card.name,
    quantity: entry.quantity,
    ...(isCommander ? { isCommander: true } : {}),
  }));
}

export async function fetchMoxfieldDeck(deckId: string): Promise<ParsedDeck> {
  const url = `${MOXFIELD_API_BASE}/decks/all/${deckId}`;

  const response = await moxfieldFetch(url);

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Deck not found: ${deckId}`);
    }
    if (response.status === 429) {
      throw new Error('Moxfield rate limit exceeded. Please try again later.');
    }
    throw new Error(
      `Failed to fetch Moxfield deck: ${response.status} ${response.statusText}`
    );
  }

  const data: MoxfieldDeckResponse = await response.json();

  const commanders = [
    ...boardCards(data.boards.commanders, true),
    ...boardCards(data.boards.companions, true),
  ];
  const mainboard = boardCards(data.boards.mainboard, false);

  return {
    name: data.name,
    commanders,
    mainboard,
  };
}

export async function fetchDeckFromMoxfieldUrl(url: string): Promise<ParsedDeck> {
  const deckId = extractMoxfieldDeckId(url);
  if (!deckId) {
    throw new Error(`Invalid Moxfield URL: ${url}`);
  }
  return fetchMoxfieldDeck(deckId);
}

File: worker_flutter/lib/ingestion/moxfield.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'parsed_deck.dart';

/// Moxfield deck fetcher. Port of `api/lib/ingestion/moxfield.ts` plus
/// a built-in 1-req/sec rate limiter (the TS version lives in
/// `moxfield-service.ts`).
///
/// Moxfield requires every API caller to identify themselves via the
/// User-Agent header. We default to `MagicBracketSimulator-Worker/0.1`
/// but accept an override so users (or test stubs) can supply their
/// own.
class MoxfieldClient {
  MoxfieldClient({http.Client? client, String? userAgent})
    : _http = client ?? http.Client(),
      _userAgent = userAgent ?? 'MagicBracketSimulator-Worker/0.1';

  final http.Client _http;
  final String _userAgent;
  static const _kBase = 'https://api2.moxfield.com/v3';
  static const _kMinIntervalMs = 1000;
  static DateTime _lastRequest = DateTime.fromMillisecondsSinceEpoch(0);

  Future<ParsedDeck> fetchByUrl(String url) async {
    final id = extractMoxfieldDeckId(url);
    if (id == null) throw FormatException('Invalid Moxfield URL: $url');
    return fetchByDeckId(id);
  }

  Future<ParsedDeck> fetchByDeckId(String deckId) async {
    await _throttle();
    final uri = Uri.parse('$_kBase/decks/all/$deckId');
    final resp = await _http.get(
      uri,
      headers: {'Accept': 'application/json', 'User-Agent': _userAgent},
    );
    if (resp.statusCode != 200) {
      if (resp.statusCode == 404) {
        throw StateError('Deck not found: $deckId');
      }
      if (resp.statusCode == 429) {
        throw StateError('Moxfield rate limit exceeded. Try again later.');
      }
      throw StateError(
        'Failed to fetch Moxfield deck: HTTP ${resp.statusCode}',
      );
    }
    final data = json.decode(resp.body) as Map<String, dynamic>;
    final boards =
        (data['boards'] as Map?)?.cast<String, dynamic>() ?? const {};

    final commanders = <DeckCard>[
      ..._boardCards(boards['commanders'], isCommander: true),
      ..._boardCards(boards['companions'], isCommander: true),
    ];
    final mainboard = _boardCards(boards['mainboard'], isCommander: false);

    return ParsedDeck(
      name: (data['name'] as String?) ?? 'Imported Moxfield Deck',
      commanders: commanders,
      mainboard: mainboard,
    );
  }

  Future<void> _throttle() async {
    final elapsed = DateTime.now().difference(_lastRequest).inMilliseconds;
    if (elapsed < _kMinIntervalMs) {
      await Future<void>.delayed(
        Duration(milliseconds: _kMinIntervalMs - elapsed),
      );
    }
    _lastRequest = DateTime.now();
  }

  List<DeckCard> _boardCards(dynamic raw, {required bool isCommander}) {
    if (raw is! Map) return const [];
    final cards = raw['cards'];
    if (cards is! Map) return const [];
    return cards.values
        .whereType<Map>()
        .map((entry) {
          final card = (entry['card'] as Map?) ?? const {};
          return DeckCard(
            name: (card['name'] as String?) ?? '',
            quantity: (entry['quantity'] as num?)?.toInt() ?? 1,
            isCommander: isCommander,
          );
        })
        .where((c) => c.name.isNotEmpty)
        .toList(growable: false);
  }
}

bool isMoxfieldUrl(String url) {
  final p = Uri.tryParse(url);
  if (p == null) return false;
  if (p.scheme != 'http' && p.scheme != 'https') return false;
  if (p.host != 'moxfield.com' && p.host != 'www.moxfield.com') return false;
  return RegExp(r'^/decks/([a-zA-Z0-9_-]+)').hasMatch(p.path);
}

String? extractMoxfieldDeckId(String url) {
  final p = Uri.tryParse(url);
  if (p == null) return null;
  if (p.scheme != 'http' && p.scheme != 'https') return null;
  if (p.host != 'moxfield.com' && p.host != 'www.moxfield.com') return null;
  final m = RegExp(r'^/decks/([a-zA-Z0-9_-]+)').firstMatch(p.path);
  return m?.group(1);
}
